File: classes_without_diagrams/library_classes.py
class Library:
    def __init__(self, name, address):
        self.name = name
        self.address = address
        self.books = []
        self.members = []

    def add_book(self, book):
        self.books.append(book)
        print(f"Book '{book.title}' added to the library.")

    def register_member(self, member):
        self.members.append(member)
        print(f"Member '{member.name}' registered to the library.")


class Book:
    def __init__(self, title, author, isbn):
        self.title = title
        self.author = author
        self.isbn = isbn
        self.borrower = None

    def borrow(self, member):
        if self.borrower is None:
            self.borrower = member
            print(f"'{self.title}' is borrowed by {member.name}.")
        else:
            print(f"'{self.title}' is already borrowed by {self.borrower.name}.")

    def return_book(self):
        if self.borrower:
            print(f"'{self.title}' is returned by {self.borrower.name}.")
            self.borrower = None
        else:
            print(f"'{self.title}' is not currently borrowed.")


class Member:
    def __init__(self, name, membership_id):
        self.name = name
        self.membership_id = membership_id
        self.borrowed_books = []

    def borrow_book(self, book):
        if len(self.borrowed_books) < 5:
            book.borrow(self)
            self.borrowed_books.append(book)
        else:
            print(f"{self.name} has reached the borrowing limit.")

    def return_book(self, book):
        if book in self.borrowed_books:
            book.return_book()
            self.borrowed_books.remove(book)
        else:
            print(f"{self.name} doesn't have the book '{book.title}' to return.")


class Librarian:
    def __init__(self, name, employee_id):
        self.name = name
        self.employee_id = employee_id

    def catalog_book(self, library, book):
        library.add_book(book)
        print(f"Librarian '{self.name}' cataloged '{book.title}'.")

    def enroll_member(self, library, member):
        library.register_member(member)
        print(f"Librarian '{self.name}' enrolled member '{member.name}'.")

File: classes_without_diagrams/library_classes.php
<?php

class Library {
    private $name;
    private $address;
    private $books = [];
    private $members = [];

    public function __construct($name, $address) {
        $this->name = $name;
        $this->address = $address;
    }

    public function addBook($book) {
        $this->books[] = $book;
        echo "Book '{$book->getTitle()}' added to the library.\n";
    }

    public function registerMember($member) {
        $this->members[] = $member;
        echo "Member '{$member->getName()}' registered to the library.\n";
    }
}

class Book {
    private $title;
    private $author;
    private $isbn;
    private $borrower = null;

    public function __construct($title, $author, $isbn) {
        $this->title = $title;
        $this->author = $author;
        $this->isbn = $isbn;
    }

    public function borrow($member) {
        if ($this->borrower === null) {
            $this->borrower = $member;
            echo "'{$this->title}' is borrowed by {$member->getName()}.\n";
        } else {
            echo "'{$this->title}' is already borrowed by {$this->borrower->getName()}.\n";
        }
    }

    public function returnBook() {
        if ($this->borrower !== null) {
            echo "'{$this->title}' is returned by {$this->borrower->getName()}.\n";
            $this->borrower = null;
        } else {
            echo "'{$this->title}' is not currently borrowed.\n";
        }
    }

    public function getTitle() {
        return $this->title;
    }
}

class Member {
    private $name;
    private $membershipId;
    private $borrowedBooks = [];

    public function __construct($name, $membershipId) {
        $this->name = $name;
        $this->membershipId = $membershipId;
    }

    public function borrowBook($book) {
        if (count($this->borrowedBooks) < 5) {
            $book->borrow($this);
            $this->borrowedBooks[] = $book;
        } else {
            echo "{$this->name} has reached the borrowing limit.\n";
        }
    }

    public function returnBook($book) {
        $key = array_search($book, $this->borrowedBooks, true);
        if ($key !== false) {
            $book->returnBook();
            unset($this->borrowedBooks[$key]);
        } else {
            echo "{$this->name} doesn't have the book '{$book->getTitle()}' to return.\n";
        }
    }

    public function getName() {
        return $this->name;
    }
}

class Librarian {
    private $name;
    private $employeeId;

    public function __construct($name, $employeeId) {
        $this->name = $name;
        $this->employeeId = $employeeId;
    }

    public function catalogBook($library, $book) {
        $library->addBook($book);
        echo "Librarian '{$this->name}' cataloged '{$book->getTitle()}'.\n";
    }

    public function enrollMember($library, $member) {
        $library->registerMember($member);
        echo "Librarian '{$this->name}' enrolled member '{$member->getName()}'.\n";
    }
}
